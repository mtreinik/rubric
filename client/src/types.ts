export type MultiSelectCriterionType = {
  title: string
  options: string[]
}

export type TextAreaCriterionType = {
  title: string
}

export type InfoCriterionType = {
  title: string
}

export type SliderRowType = {
  title: string
  value: number
}

export type SliderCriterionType = {
  title: string
  options: string[]
  rows: SliderRowType[]
}

export type CriterionType =
  | MultiSelectCriterionType
  | TextAreaCriterionType
  | InfoCriterionType
  | SliderCriterionType

export interface CriterionAndType {
  type: string
  criterion: CriterionType
}

export interface CriterionWithOptionsAndType {
  type: 'MultiSelectCriterion' | 'SliderCriterion'
  criterion: MultiSelectCriterionType | SliderCriterionType
}

export function isCriterionWithOptionsAndType(
  criterionAndType: CriterionAndType
): criterionAndType is CriterionWithOptionsAndType {
  return (
    criterionAndType.type === 'MultiSelectCriterion' ||
    criterionAndType.type === 'SliderCriterion'
  )
}

export interface SliderCriterionAndType {
  type: 'SliderCriterion'
  criterion: SliderCriterionType
}

export function isSliderCriterionAndType(
  criterionAndType: CriterionAndType
): criterionAndType is SliderCriterionAndType {
  return criterionAndType.type === 'SliderCriterion'
}

export interface SectionType {
  title: string
  criterions: CriterionAndType[]
}

export type SelectionType = 'select' | 'deselect' | undefined

export type AppState = {
  sections: SectionType[]
  selection: SelectionType
  language: string
  showRubricEditor: boolean
  version: number
  dirty: boolean
}

export type LanguageCode = 'fi' | 'en'
export type Language = {
  code: LanguageCode
  name: string
}
