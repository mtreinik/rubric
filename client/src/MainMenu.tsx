import React, { ChangeEvent, MutableRefObject, useRef, useState } from 'react'
import {
  Button,
  Divider,
  Grid,
  Icon,
  ListItemIcon,
  ListItemText,
  Menu,
  MenuItem,
  Snackbar,
} from '@material-ui/core'
import Alert from '@material-ui/lab/Alert'
import { AppState, SectionType, Language, LanguageCode } from './types'
import { TFunction } from 'i18next'
import Ajv from 'ajv'
import rubricSchema from './rubricSchema.json'
import AlertDialog, { AlertDialogStateType } from './AlertDialog'
import DownloadDialog from './DownloadDialog'
import AcceptDialog from './AcceptDialog'

const ajv = new Ajv()

interface Props {
  appState: AppState
  language: string
  setSectionsAndCleanAppState: (sections: SectionType[]) => void
  toggleRubricEditor: () => void
  changeLanguage: (language: LanguageCode) => void
  cleanAppState: () => void
  t: TFunction
}

const languages: Language[] = [
  { code: 'fi', name: 'suomi' },
  { code: 'en', name: 'English' },
]

const closedAlertDialog: AlertDialogStateType = {
  open: false,
  title: '',
  message: '',
  details: '',
}

const MainMenu = (props: Props): JSX.Element => {
  const [anchorEl, setAnchorEl] = useState<HTMLElement | null>(null)
  const [alertDialogState, setAlertDialogState] = useState(closedAlertDialog)
  const [snackbarOpen, setSnackbarOpen] = useState(false)
  const [downloadDialogOpen, setDownloadDialogOpen] = useState(false)
  const [acceptNewDialogOpen, setAcceptNewDialogOpen] = useState(false)
  const [acceptOpenDialogOpen, setAcceptOpenDialogOpen] = useState(false)

  const t = props.t
  const uploaderRefObject = useRef() as MutableRefObject<HTMLInputElement>

  const openAlertDialog = (title: string, message: string, details: string) => {
    setAlertDialogState({ open: true, title, message, details })
  }

  const closeAlertDialog = () => {
    setAlertDialogState(closedAlertDialog)
  }

  const warnAndCloseMenu = (message: string, details = ''): void => {
    console.warn('Error opening rubric', message, details)
    openAlertDialog(t('errorOpeningRubric'), message, details.toString())
    setAnchorEl(null)
  }

  const uploadRubric = (event: ChangeEvent<HTMLInputElement>): void => {
    if (!event.target.files) {
      warnAndCloseMenu(t('noFileWasSelected'))
      return
    }
    const file = event.target.files.item(0)
    if (file) {
      const reader = new FileReader()
      reader.onload = (e) => {
        if (!e || !e.target || !e.target.result) {
          warnAndCloseMenu(t('couldNotReadFile'))
          return
        }

        const data = e.target.result
        if (typeof data !== 'string') {
          warnAndCloseMenu(t('fileHasInvalidType'))
          return
        }

        try {
          const json = JSON.parse(data)
          const isValid = ajv.validate(rubricSchema, json)
          if (!isValid) {
            warnAndCloseMenu(t('invalidFileFormat'), ajv.errorsText())
            return
          }

          const sections = JSON.parse(data)
          props.setSectionsAndCleanAppState(sections)
          setAnchorEl(null)
          setSnackbarOpen(true)
        } catch (error) {
          let errorMessage = 'Unknown error'

          if (error instanceof Error) {
            errorMessage = error.message
          } else if (typeof error === 'string') {
            errorMessage = error
          }
          warnAndCloseMenu(t('invalidJSONFile'), errorMessage)
        }
      }
      reader.onerror = () => {
        warnAndCloseMenu(t('openingWasAborted'))
      }
      reader.readAsText(file)
    }
  }
  const createNewRubric = () => props.setSectionsAndCleanAppState([])

  return (
    <Grid container spacing={4} alignItems="center">
      <Grid item>
        <Button
          aria-owns={anchorEl ? 'simple-menu' : undefined}
          onClick={(event: React.MouseEvent<HTMLElement>) =>
            setAnchorEl(event.currentTarget)
          }
          startIcon={<Icon>menu</Icon>}
        >
          {t('menu')}
        </Button>
        <Menu
          id="simple-menu"
          anchorEl={anchorEl}
          open={Boolean(anchorEl)}
          onClose={() => setAnchorEl(null)}
        >
          <MenuItem
            onClick={() => {
              if (props.appState.dirty) {
                setAcceptOpenDialogOpen(true)
              } else {
                uploaderRefObject.current.click()
              }
            }}
          >
            <ListItemIcon>
              <Icon>open_in_browser</Icon>
            </ListItemIcon>
            <ListItemText>{t('openRubric')}</ListItemText>
          </MenuItem>

          <MenuItem onClick={() => setDownloadDialogOpen(true)}>
            <ListItemIcon>
              <Icon>save_alt</Icon>
            </ListItemIcon>
            <ListItemText>{t('saveRubricEllipsis')}</ListItemText>
          </MenuItem>

          <MenuItem
            onClick={() => {
              if (props.appState.dirty) {
                setAcceptNewDialogOpen(true)
              } else {
                createNewRubric()
              }
              setAnchorEl(null)
            }}
          >
            <ListItemIcon>
              <Icon>add</Icon>
            </ListItemIcon>
            <ListItemText>{t('newRubric')}</ListItemText>
          </MenuItem>

          <Divider />

          <MenuItem
            onClick={() => {
              setAnchorEl(null)
              props.toggleRubricEditor()
            }}
          >
            <ListItemIcon>
              <Icon>
                {props.appState.showRubricEditor ? (
                  <Icon>close</Icon>
                ) : (
                  <Icon>edit</Icon>
                )}
              </Icon>
            </ListItemIcon>
            <ListItemText>
              {props.appState.showRubricEditor
                ? t('closeRubricEditor')
                : t('editRubric')}
            </ListItemText>
          </MenuItem>

          <Divider />

          <MenuItem disabled>
            <ListItemIcon>
              <Icon>language</Icon>
            </ListItemIcon>
            <ListItemText>{t('changeLanguage')}</ListItemText>
          </MenuItem>
          {languages.map((language) => (
            <MenuItem
              onClick={() => {
                props.changeLanguage(language.code)
                setAnchorEl(null)
              }}
              key={'language-' + language.code}
            >
              <ListItemIcon>
                {language.code}
                {language.code === props.language && <Icon>check</Icon>}
              </ListItemIcon>
              <ListItemText>{language.name}</ListItemText>
            </MenuItem>
          ))}
        </Menu>
        <DownloadDialog
          open={downloadDialogOpen}
          sections={props.appState.sections}
          closeDialogOnCancel={() => {
            setDownloadDialogOpen(false)
            setAnchorEl(null)
          }}
          closeDialogOnSave={() => {
            setDownloadDialogOpen(false)
            setAnchorEl(null)
            props.cleanAppState()
          }}
          t={t}
        />
        <AlertDialog
          dialogState={alertDialogState}
          closeDialog={closeAlertDialog}
          t={t}
        />
        <AcceptDialog
          open={acceptNewDialogOpen}
          title={t('discardChanges')}
          message={t('discardUnsavedChangesAndCreateNew')}
          cancelButtonKey="cancel"
          acceptButtonKey="discardAndCreateNew"
          closeDialogOnCancel={() => {
            setAcceptNewDialogOpen(false)
            setAnchorEl(null)
          }}
          closeDialogOnAccept={() => {
            setAcceptNewDialogOpen(false)
            setAnchorEl(null)
            createNewRubric()
          }}
          t={t}
        />
        <AcceptDialog
          open={acceptOpenDialogOpen}
          title={t('discardChanges')}
          message={t('discardUnsavedChangesAndOpen')}
          cancelButtonKey="cancel"
          acceptButtonKey="discardAndOpen"
          closeDialogOnCancel={() => {
            setAcceptOpenDialogOpen(false)
            setAnchorEl(null)
          }}
          closeDialogOnAccept={() => {
            setAcceptOpenDialogOpen(false)
            setAnchorEl(null)
            uploaderRefObject.current.click()
          }}
          t={t}
        />
        <input
          type="file"
          accept="text/json,.json"
          multiple={false}
          ref={uploaderRefObject}
          style={{ display: 'none' }}
          onChange={uploadRubric}
          onClick={(event: React.MouseEvent<HTMLInputElement>) =>
            (event.currentTarget.value = '')
          }
        />
        <Snackbar
          anchorOrigin={{
            vertical: 'top',
            horizontal: 'center',
          }}
          open={snackbarOpen}
          autoHideDuration={4000}
          onClose={() => setSnackbarOpen(false)}
        >
          <Alert severity="success">{t('openedRubric')}</Alert>
        </Snackbar>
      </Grid>
      {props.appState.dirty && (
        <Grid item>
          <Button
            onClick={() => setDownloadDialogOpen(true)}
            startIcon={<Icon>save_alt</Icon>}
          >
            {t('saveChanges')}
          </Button>
        </Grid>
      )}
    </Grid>
  )
}

export default MainMenu
