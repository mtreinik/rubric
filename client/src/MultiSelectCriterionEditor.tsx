import React from 'react'
import { MultiSelectCriterionType } from './types'
import { Button, Grid, Icon, IconButton, TextField } from '@material-ui/core'
import { TFunction } from 'i18next'

interface Props {
  criterion: MultiSelectCriterionType
  criterionIndex: number
  removeCriterion: (criterionIndex: number) => () => void
  editCriterion: (criterionTitle: string) => void
  addOption: () => void
  removeOption: (optionIndex: number) => void
  editOption: (optionIndex: number) => (optionTitle: string) => void
  t: TFunction
}

const MultiSelectCriterionEditor = (props: Props): JSX.Element => {
  const t = props.t
  const handleCriterionTitleChange = (
    event: React.ChangeEvent<HTMLInputElement>
  ): void => {
    props.editCriterion(event.target.value)
  }
  const handleOptionTitleChange =
    (optionIndex: number) =>
    (event: React.ChangeEvent<HTMLInputElement>): void => {
      props.editOption(optionIndex)(event.target.value)
    }

  const clickables = props.criterion.options.map((option, optionIndex) => (
    <Grid item key={'option-' + optionIndex}>
      <Grid container>
        <Grid item xs={2}>
          <IconButton
            onClick={() => props.removeOption(optionIndex)}
            color="secondary"
            size="small"
          >
            <Icon fontSize="small">remove_circle</Icon>
          </IconButton>
        </Grid>
        <Grid item xs={10}>
          <TextField
            value={option}
            onChange={handleOptionTitleChange(optionIndex)}
            autoFocus
            fullWidth
          />
        </Grid>
      </Grid>
    </Grid>
  ))
  return (
    <Grid container direction="row" spacing={2}>
      <Grid item xs={6}>
        <TextField
          value={props.criterion.title}
          helperText={t('multiSelectTitleHelperText')}
          onChange={handleCriterionTitleChange}
          fullWidth
          autoFocus
        />
      </Grid>
      <Grid item xs={6}>
        <Grid container direction="column" spacing={2}>
          {clickables}
          <Grid item>
            <Button
              style={{ marginLeft: '0.1em' }}
              onClick={props.addOption}
              size="small"
              startIcon={<Icon>add_circle</Icon>}
            >
              {t('addOption')}
            </Button>
          </Grid>
        </Grid>
      </Grid>
    </Grid>
  )
}

export default MultiSelectCriterionEditor
