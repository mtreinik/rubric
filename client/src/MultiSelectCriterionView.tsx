import React, { useState } from 'react'
import { createSeparatedReactNodes } from './react-utils'
import Clickable from './Clickable'
import { SelectionType } from './types'
import { Snackbar } from '@material-ui/core'
import Alert from '@material-ui/lab/Alert'
import { TFunction } from 'i18next'

interface Props {
  title: string
  options: string[]
  selection: SelectionType
  t: TFunction
}

const MultiSelectCriterionView = (props: Props): JSX.Element => {
  const [snackbarOpen, setSnackbarOpen] = useState(false)
  const t = props.t
  const clickables = createSeparatedReactNodes(
    props.options,
    (value, index) => (
      <Clickable
        key={'clickable-' + index}
        value={value}
        selection={props.selection}
        warnAboutSelection={() => setSnackbarOpen(true)}
      />
    ),
    (index) => <span key={'separator' + index}> / </span>
  )
  return (
    <div className="criterion">
      <span className="multiSelectCriterionTitle">{props.title}</span>{' '}
      {clickables}
      <Snackbar
        anchorOrigin={{
          vertical: 'top',
          horizontal: 'center',
        }}
        open={snackbarOpen}
        autoHideDuration={4000}
        onClose={() => setSnackbarOpen(false)}
      >
        <Alert severity="info">{t('deselectBeforeUsingMultiSelect')}</Alert>
      </Snackbar>
    </div>
  )
}

export default MultiSelectCriterionView
