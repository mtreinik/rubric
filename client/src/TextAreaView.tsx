import React, { useState } from 'react'
import { SelectionType } from './types'
import DOMPurify from 'dompurify'

interface Props {
  title: string
  selection?: SelectionType
  value?: string
}

const TextAreaView = (props: Props): JSX.Element => {
  const [value, setValue] = useState(props.value || '')

  const handleValueChange = (
    event: React.ChangeEvent<HTMLTextAreaElement>
  ): void => {
    setValue(event.target.value)
  }

  const escapeHtml = (unsafe: string) =>
    unsafe
      .replace(/&/g, '&amp;')
      .replace(/</g, '&lt;')
      .replace(/>/g, '&gt;')
      .replace(/"/g, '&quot;')
      .replace(/'/g, '&#039;')

  const replaceNewlinesAndEscapeHtml = (value: string) =>
    DOMPurify.sanitize(escapeHtml(value).replace(/\n/g, '<br />'))

  const valueComponent =
    props.selection === 'select' ? (
      <span
        dangerouslySetInnerHTML={{
          __html: replaceNewlinesAndEscapeHtml(value),
        }}
      />
    ) : (
      <textarea
        rows={4}
        cols={80}
        value={value}
        onChange={handleValueChange}
        className="textArea"
      />
    )

  if (props.selection === 'select' && !value) {
    return <span></span>
  } else {
    return (
      <div
        className={
          props.selection === 'select'
            ? 'textAreaCriterionCell withBorder'
            : 'textAreaCriterionCell'
        }
      >
        <div className="criterionTitle">{props.title}</div>
        <div>{valueComponent}</div>
      </div>
    )
  }
}

export default TextAreaView
