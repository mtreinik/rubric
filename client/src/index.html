<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta
      name="viewport"
      content="minimum-scale=1, initial-scale=1, width=device-width, shrink-to-fit=no"
    />
    <link
      rel="stylesheet"
      href="https://fonts.googleapis.com/icon?family=Material+Icons"
    />
    <link
      rel="stylesheet"
      href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap"
    />
    <script>
      window.addEventListener('beforeunload', function (event) {
        event.preventDefault()
        event.returnValue = '' // Required for Chrome
      })
    </script>
    <title>Rubric</title>
  </head>
  <body>
    <noscript>This app requires JavaScript</noscript>
    <div id="root"></div>
  </body>
</html>
