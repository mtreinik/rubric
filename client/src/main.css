#root {
  margin: 1em;
}

.rubricViewPanel {
  padding-left: 1em;
}

#content {
  padding-bottom: 1em;
  font-family: Times, 'Times New Roman', Serif;
}

.rubricTitle {
  font-size: x-large;
}

.sectionTitle {
  font-weight: bold;
  padding-bottom: 0.5em;
}

.criterionTitle {
  font-weight: bold;
  padding-right: 1em;
}

.multiSelectCriterionTitle {
  font-weight: bold;
}

.criterion,
.infoCriterion,
.textAreaCriterion,
.sliderTable {
  padding-left: 2em;
  padding-bottom: 0.4em;
}

.textAreaCriterionCell .withBorder {
  border: 1px solid black;
}

.clickable {
  font-family: Times, 'Times New Roman', Serif;
}

.textArea {
  font-family: Times, 'Times New Roman', Serif;
  font-size: x-small;
}

.sliderTable td,
.sliderTable th {
  border: 1px solid black;
}

.sliderHeaderRow {
  font-size: 0.5em;
}

.sliderHeaderHeader {
  font-size: 0.5em;
  padding: 0em;
  width: 90em;
}

.sliderHeaderCell {
  text-align: center;
  font-size: 0.6em;
  padding: 0em;
  width: 11em;
}

.sliderRowHeader {
  font-size: 0.8em;
  line-height: 0.8;
  padding: 0.1em;
}

.sliderColumn {
  padding: 0em;
  padding-left: 0.2em;
}

.sliderCell {
  font-size: 1em;
  padding: 0em;
  cursor: pointer;
  vertical-align: middle;
}

button.clickable {
  border: 1px solid #ffffff;
  padding: 4px 6px 4px 6px;
  background-color: #ffffff;
  border-radius: 4px;
}

.clickable.full {
  background-color: #80ff80;
  font-weight: bold;
}

.clickable.partial {
  background-color: #ffff40;
  font-weight: normal;
}

button:hover.clickable {
  border: 1px solid #808080;
}

span.sliderCell {
  border: 1px solid #ffffff;
  border-radius: 4px;
}
span:hover.sliderCell {
  border: 1px solid #808080;
}

.noselect {
  -webkit-touch-callout: none;
  -webkit-user-select: none;
  -khtml-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}
