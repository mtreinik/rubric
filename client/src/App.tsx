import React, { useEffect, useState } from 'react'
import { Button, Icon, CssBaseline, Grid } from '@material-ui/core/'
import { createMuiTheme, ThemeProvider } from '@material-ui/core/styles'
import RubricEditor from './RubricEditor'
import RubricView from './RubricView'
import * as O from 'optics-ts'
import {
  AppState,
  isSliderCriterionAndType,
  LanguageCode,
  SectionType,
  SelectionType,
} from './types'
import { useTranslation } from 'react-i18next'
import i18n from './i18n'
import { swapElements } from './array-utils'
import MainMenu from './MainMenu'

const emptyAppState: AppState = {
  sections: [],
  selection: undefined,
  language: '',
  showRubricEditor: false,
  version: 1,
  dirty: false,
}

const sectionsLens = O.optic<AppState>().prop('sections')
const languageLens = O.optic<AppState>().prop('language')
const showRubricEditorLens = O.optic<AppState>().prop('showRubricEditor')
const versionLens = O.optic<AppState>().prop('version')
const dirtyLens = O.optic<AppState>().prop('dirty')
const selectionLens = O.optic<AppState>().prop('selection')

const sectionPrism = (sectionIndex: number) => sectionsLens.index(sectionIndex)

const newSectionSetter = O.optic<AppState>().prop('sections').appendTo()

const sliderRowValuesTraversal = O.optic<AppState>()
  .prop('sections')
  .elems()
  .prop('criterions')
  .elems()
  .guard(isSliderCriterionAndType)
  .prop('criterion')
  .prop('rows')
  .elems()
  .prop('value')

const theme = createMuiTheme({
  palette: {
    primary: { main: '#2e7d32' },
    secondary: { main: '#ff1744' },
    background: { default: '#ffffff' },
  },
})

const setSelectionEnabled = (enabled: boolean) => {
  const elementId = 'content'
  const element = document.getElementById(elementId)
  if (!element) {
    console.error('could not get element', elementId)
    return
  }
  element.className = enabled ? '' : 'noselect'
}

const deselectAll = () => {
  const selection = window.getSelection()
  if (!selection) {
    return
  }
  selection.removeAllRanges()
  setSelectionEnabled(false)
}

const selectElement = (elementId: string) => () => {
  const element = document.getElementById(elementId)
  if (!element) {
    console.error('could not get element', elementId)
    return
  }
  const range = document.createRange()
  range.selectNodeContents(element)
  const selection = window.getSelection()
  if (!selection) {
    return
  }
  selection.removeAllRanges()
  setSelectionEnabled(true)
  selection.addRange(range)
  document.execCommand('copy')
}

const App = (): JSX.Element => {
  const [appState, setAppState] = useState(emptyAppState)
  const { t } = useTranslation()

  React.useEffect(() => {
    setSelectionEnabled(false)
  })

  const cleanAppState = (): void => {
    setAppState(O.set(dirtyLens)(false)(appState))
  }

  const changeAppState = (newAppState: AppState): void => {
    if (appState.dirty) {
      setAppState(newAppState)
    } else {
      setAppState(O.set(dirtyLens)(true)(newAppState))
    }
  }

  const addSection = (): void => {
    changeAppState(
      O.set(newSectionSetter)({
        title: '',
        criterions: [],
      })(appState)
    )
  }

  const removeSection = (sectionIndex: number): void => {
    if (sectionIndex >= appState.sections.length - 1) {
      console.error(`Cannot remove section at ${sectionIndex}`)
    }
    changeAppState(O.remove(sectionPrism(sectionIndex))(appState))
  }

  const editSection = (sectionIndex: number) => (section: SectionType) => {
    changeAppState(O.set(sectionPrism(sectionIndex))(section)(appState))
  }

  const moveSectionUp = (sectionIndex: number): void => {
    if (sectionIndex < 1) {
      console.error(`Cannot move up section at ${sectionIndex}`)
      return
    }
    moveSectionDown(sectionIndex - 1)
  }

  const moveSectionDown = (sectionIndex: number): void => {
    if (sectionIndex >= appState.sections.length) {
      console.error(`Cannot move down section at ${sectionIndex}`)
      return
    }
    const newSections = swapElements(sectionIndex, appState.sections)
    changeAppState(O.set(sectionsLens)(newSections)(appState))
  }

  const setSelection = (selection: SelectionType) => {
    setAppState(O.set(selectionLens)(selection)(appState))
  }

  useEffect(() => {
    if (appState.selection === 'deselect') {
      deselectAll()
      setSelection(undefined)
    } else if (appState.selection === 'select') {
      selectElement('content')()
    }
  }, [appState.selection])

  useEffect(() => {
    document.title = t('appTagline')
  }, [appState.language])

  const changeLanguage = (language: LanguageCode): void => {
    i18n.changeLanguage(language)
    setAppState(O.set(languageLens)(language)(appState))
  }

  const toggleRubricEditor = (): void => {
    setAppState(
      O.set(showRubricEditorLens)(!appState.showRubricEditor)(appState)
    )
  }

  const setSliderRowValue =
    (sectionIndex: number, criterionIndex: number) =>
    (rowIndex: number, value: number): void => {
      const valuePrism = O.optic<AppState>()
        .prop('sections')
        .index(sectionIndex)
        .prop('criterions')
        .index(criterionIndex)
        .guard(isSliderCriterionAndType)
        .prop('criterion')
        .prop('rows')
        .index(rowIndex)
        .prop('value')
      setAppState(O.set(valuePrism)(value)(appState))
    }

  const setSectionsAndCleanAppState = (sections: SectionType[]): void => {
    setAppState(
      O.pipe(
        appState,
        O.set(selectionLens)('deselect' as const),
        O.set(dirtyLens)(false),
        O.set(sectionsLens)(sections)
      )
    )
  }

  const reset = (): void => {
    setAppState(
      O.pipe(
        appState,
        O.set(versionLens)(appState.version + 1),
        O.set(selectionLens)('deselect' as const),
        O.set(sliderRowValuesTraversal)(-1)
      )
    )
  }

  return (
    <React.Fragment>
      <CssBaseline />
      <ThemeProvider theme={theme}>
        <CssBaseline />
        <Grid container direction="row" spacing={2}>
          <Grid
            item
            xs={appState.showRubricEditor ? 6 : 1}
            style={{
              transitionDuration: '0.5s',
              transitionProperty: 'flex-basis width',
            }}
          >
            <RubricEditor
              appState={appState}
              addSection={addSection}
              removeSection={removeSection}
              editSection={editSection}
              moveSectionUp={moveSectionUp}
              moveSectionDown={moveSectionDown}
              t={t}
              toggleRubricEditor={toggleRubricEditor}
            />
          </Grid>
          <Grid
            item
            xs={appState.showRubricEditor ? 6 : 11}
            style={{
              transitionDuration: '0.5s',
              transitionProperty: 'flex-basis width',
            }}
          >
            <Grid container spacing={2} className="rubricViewPanel">
              <Grid item xs={12}>
                <MainMenu
                  appState={appState}
                  language={i18n.language}
                  toggleRubricEditor={toggleRubricEditor}
                  setSectionsAndCleanAppState={setSectionsAndCleanAppState}
                  changeLanguage={changeLanguage}
                  cleanAppState={cleanAppState}
                  t={t}
                />
              </Grid>
              <Grid item xs={12}>
                <div id="content">
                  <RubricView
                    sections={appState.sections}
                    selection={appState.selection}
                    version={appState.version}
                    setSliderRowValue={setSliderRowValue}
                    t={t}
                  />
                </div>
              </Grid>
              <Grid item xs={12}>
                <Grid container>
                  {!appState.selection && (
                    <Grid item xs={6}>
                      <Button
                        onClick={() => setSelection('select')}
                        variant="contained"
                        color="primary"
                        startIcon={<Icon>content_copy</Icon>}
                      >
                        {t('selectAndCopy')}
                      </Button>
                    </Grid>
                  )}
                  {appState.selection && (
                    <Grid item xs={6}>
                      <Button
                        onClick={() => setSelection('deselect')}
                        variant="contained"
                        color="primary"
                        startIcon={<Icon>clear</Icon>}
                      >
                        {t('deselect')}
                      </Button>
                    </Grid>
                  )}
                  <Grid item xs={6}>
                    <Button
                      onClick={reset}
                      variant="outlined"
                      color="secondary"
                      startIcon={<Icon>delete_forever</Icon>}
                    >
                      {t('reset')}
                    </Button>
                  </Grid>
                </Grid>
              </Grid>
            </Grid>
          </Grid>
        </Grid>
      </ThemeProvider>
    </React.Fragment>
  )
}

export default App
