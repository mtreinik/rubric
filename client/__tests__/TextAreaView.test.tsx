import React from 'react'
import { mount, shallow } from 'enzyme'
import TextAreaView from '../src/TextAreaView'
import { SelectionType } from '../src/types'

describe('TextAreaView', () => {
  describe('renders correctly when', () => {
    const expectValidComponent = (
      title: string,
      value: string,
      selection?: SelectionType
    ): void => {
      const component = shallow(
        <TextAreaView title={title} value={value} selection={selection} />
      )
      expect(component).toMatchSnapshot()
    }

    describe('no selection and', () => {
      it('empty title and empty value', () => {
        expectValidComponent('', '')
      })
      it('text title and empty value', () => {
        expectValidComponent('xyzzy', '')
      })
      it('text title and text value', () => {
        expectValidComponent('foo', 'bar')
      })
      it('text title and value with newlines', () => {
        expectValidComponent('foo', 'bar\n\nbaz')
      })
      it('text title and value with XSS attack', () => {
        expectValidComponent(
          'title',
          '<span onmouseover="alert(document.cookie)">this is an\nXSS attack</span>'
        )
      })
      it('editing the value', () => {
        const component = mount(<TextAreaView title="title" />)
        expect(component.find('textarea').props().value).toEqual('')
        component
          .find('textarea')
          .simulate('change', { target: { value: 'foo\n\nbar' } })
        expect(component.find('textarea').props().value).toEqual('foo\n\nbar')
      })
    })

    describe('selection and', () => {
      it('empty title and empty value', () => {
        expectValidComponent('', '', 'select')
      })
      it('text title and empty value', () => {
        expectValidComponent('xyzzy', '', 'select')
      })
      it('text title and text value', () => {
        expectValidComponent('foo', 'bar', 'select')
      })
      it('text title and value with newlines', () => {
        expectValidComponent('foo', 'bar\n\nbaz', 'select')
      })
      it('text title and value with XSS attack', () => {
        expectValidComponent(
          'title',
          '<span onmouseover="alert(document.cookie)">this is an\nXSS attack</span>',
          'select'
        )
      })
    })
  })
  describe('behaves correctly when', () => {
    it('editing the value and turning selection on/off', () => {
      const component = mount(<TextAreaView title="title" />)
      expect(component.find('textarea').props().value).toEqual('')
      expect(component.find('span')).toEqual({})

      component
        .find('textarea')
        .simulate('change', { target: { value: 'foo<b>bar</b>' } })
      expect(component.find('textarea').props().value).toEqual('foo<b>bar</b>')
      expect(component.find('span')).toEqual({})

      component
        .find('textarea')
        .simulate('change', { target: { value: 'foo\n\nbar' } })
      expect(component.find('textarea').props().value).toEqual('foo\n\nbar')
      expect(component.find('span')).toEqual({})

      component.setProps({ selection: 'select' })
      expect(component.find('textarea')).toEqual({})
      expect(component.find('span').html()).toEqual(
        '<span>foo<br><br>bar</span>'
      )

      component.setProps({ selection: '' })
      expect(component.find('textarea').props().value).toEqual('foo\n\nbar')
      expect(component.find('span')).toEqual({})
    })
  })
})
