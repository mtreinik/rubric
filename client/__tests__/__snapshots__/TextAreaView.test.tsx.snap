// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextAreaView renders correctly when no selection and empty title and empty value 1`] = `
<div
  className="textAreaCriterionCell"
>
  <br />
  <div
    className="criterionTitle"
  />
  <div>
    <textarea
      className="textArea"
      cols={80}
      onChange={[Function]}
      rows={4}
      value=""
    />
  </div>
</div>
`;

exports[`TextAreaView renders correctly when no selection and text title and empty value 1`] = `
<div
  className="textAreaCriterionCell"
>
  <br />
  <div
    className="criterionTitle"
  >
    xyzzy
  </div>
  <div>
    <textarea
      className="textArea"
      cols={80}
      onChange={[Function]}
      rows={4}
      value=""
    />
  </div>
</div>
`;

exports[`TextAreaView renders correctly when no selection and text title and text value 1`] = `
<div
  className="textAreaCriterionCell"
>
  <br />
  <div
    className="criterionTitle"
  >
    foo
  </div>
  <div>
    <textarea
      className="textArea"
      cols={80}
      onChange={[Function]}
      rows={4}
      value="bar"
    />
  </div>
</div>
`;

exports[`TextAreaView renders correctly when no selection and text title and value with XSS attack 1`] = `
<div
  className="textAreaCriterionCell"
>
  <br />
  <div
    className="criterionTitle"
  >
    title
  </div>
  <div>
    <textarea
      className="textArea"
      cols={80}
      onChange={[Function]}
      rows={4}
      value="<span onmouseover=\\"alert(document.cookie)\\">this is an
XSS attack</span>"
    />
  </div>
</div>
`;

exports[`TextAreaView renders correctly when no selection and text title and value with newlines 1`] = `
<div
  className="textAreaCriterionCell"
>
  <br />
  <div
    className="criterionTitle"
  >
    foo
  </div>
  <div>
    <textarea
      className="textArea"
      cols={80}
      onChange={[Function]}
      rows={4}
      value="bar

baz"
    />
  </div>
</div>
`;

exports[`TextAreaView renders correctly when selection and empty title and empty value 1`] = `<span />`;

exports[`TextAreaView renders correctly when selection and text title and empty value 1`] = `<span />`;

exports[`TextAreaView renders correctly when selection and text title and text value 1`] = `
<div
  className="textAreaCriterionCell withBorder"
>
  <br />
  <div
    className="criterionTitle"
  >
    foo
  </div>
  <div>
    <span
      dangerouslySetInnerHTML={
        Object {
          "__html": "bar",
        }
      }
    />
  </div>
</div>
`;

exports[`TextAreaView renders correctly when selection and text title and value with XSS attack 1`] = `
<div
  className="textAreaCriterionCell withBorder"
>
  <br />
  <div
    className="criterionTitle"
  >
    title
  </div>
  <div>
    <span
      dangerouslySetInnerHTML={
        Object {
          "__html": "&lt;span onmouseover=\\"alert(document.cookie)\\"&gt;this is an<br>XSS attack&lt;/span&gt;",
        }
      }
    />
  </div>
</div>
`;

exports[`TextAreaView renders correctly when selection and text title and value with newlines 1`] = `
<div
  className="textAreaCriterionCell withBorder"
>
  <br />
  <div
    className="criterionTitle"
  >
    foo
  </div>
  <div>
    <span
      dangerouslySetInnerHTML={
        Object {
          "__html": "bar<br><br>baz",
        }
      }
    />
  </div>
</div>
`;
